include/ginkgo/core/base/composition.hpp

```diff
@@ -16,6 +16,21 @@
 namespace gko {
 
 
+/**
+ * CompositeionBase provides the interface
+ */
+class CompositionBase {
```

Comment:
This is only necessary because we don't have a value type for Ic and Ilu, right? What if we allowed the first template parameter to be _either_ ValueType _or_ SolverType? Then the value type would always be available. Slightly more complexity in extracting the actual types to use, but no need to add a new class whose only reason for existence is only necessary temporarily.

Reply:
It is a good idea! I might need to adapt CompositionBase or different way for the mixed-IC but it should doable.
the value_type seems to be more clear way.